src/ray/object_manager/pull_manager.cc

```diff
@@ -350,24 +355,37 @@ void PullManager::TryToMakeObjectLocal(const ObjectID &object_id) {
   // sending RPCs.
   if (!request.spilled_url.empty()) {
     const auto spilled_node_id = request.spilled_node_id;
+    // Set a flag on the request so we don't send duplicate restoration requests.
+    request.is_being_restored = true;
     restore_spilled_object_(
         object_id, request.spilled_url, spilled_node_id,
-        [this, object_id, spilled_node_id](const ray::Status &status) {
+        [this, object_id, &request, spilled_node_id](const ray::Status &status) {
+          request.is_being_restored = false;
           if (!status.ok()) {
             const auto node_id_with_issue =
                 spilled_node_id.IsNil() ? self_node_id_ : spilled_node_id;
-            RAY_LOG(WARNING)
-                << "Object restoration failed and the object could "
-                   "not be "
-                   "found on any other nodes. This can happen if the location where the "
-                   "object was spilled is unreachable. This job may hang if the object "
-                   "is permanently unreachable. "
-                   "Please check the log of node of id: "
-                << node_id_with_issue << " Object id: " << object_id;
+            if (status.IsPendingRequiredData()) {
+              RAY_LOG(DEBUG) << "Object restoration failed for object " << object_id
+                             << " on node " << node_id_with_issue
+                             << " because we're waiting on pending data: "
+                             << status.ToString();
+            } else {
+              RAY_LOG(WARNING)
+                  << "Object restoration failed and the object " << object_id
+                  << " could not be found on any other nodes. This can happen "
+                     "if the location where the object was spilled is "
+                     "unreachable. This job may hang if the object is "
+                     "permanently unreachable. Please check the logs of node "
+                  << node_id_with_issue
+                  << " and the following status: " << status.ToString();
+            }
+            // Object restoration failed; we'll retry at a later time.
+            UpdateRetryTimer(request);
```

Comment:
Outside of the remote restoration requests, this also adds an exponential backoff for retrying when the object hasn't been spilled yet. I think that this is desirable, but I thought I'd call that out here in case that's controversial!